### PR TITLE
Iris & Noor

### PR DESCRIPTION
**Pull request by iris-lux**
# Assignment Submission: Rideshare Rails

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions. These questions should be answered by all team members together, not by a single teammate.

## Reflection

Prompt | Response
--- | ---
Describe the types of entity relationships you set up in your project and why you set up the relationships that way | Driver has many Trips, passenger has many trips, and Trip belongs to Passenger and Driver. We set it up that way so we'd be able to access the data and the relationship between entities easily.
Describe the role of model validations in your application | We created validations for all models. We made sure info like Vins and ratings conformed to a certain standard, and we made sure certain essential info was present (such as names for drivers/passengers). 
How did your team break up the work to be done? | Iris focused more on the back-end, such as models, and Noor focused on the front-end views. Both contributed to controllers/controller tests. 
What features did you choose to prioritize in your project, and what features, if any, did you have to set aside to meet the deadline? | We prioritized functionality over style, our stylesheets are simpler yet elegant. 
What was one thing that your team collectively gained more clarity on after completing this assignment? | How models function, and the relationships between models. 
What are two discussion points that you and your team discussed when giving/receiving feedback from each other that you would be willing to share? | Flexibility on both sides was appreciated. We had good communication around gitpush/pull which prevented confusion and code issues. 
Optional: What is the URL of your deployed Heroku app? |

### CODE REVIEW

File: app/controllers/trips_controller.rb [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+      #make a new view
+      redirect_to passenger_path(params[:passenger_id])
+    else
+      #display error message?
+      redirect_to passengers_path
+    end
+  end
+
+  def assign_rating
+    raise
```
Comment (CheezItMan) [inline app/controllers/trips_controller.rb:36]: 🙅 

```suggestion
```

File: test/controllers/trips_controller_test.rb [hunk 2]
```diff
@@ -1,23 +1,153 @@
...
+      must_respond_with :redirect
+      must_redirect_to trips_path
+    end
+
+    it "will redirect to the trips page if given an invalid id" do
+      # Your code here
+      id = -1
+      delete trip_path(id)
+      must_redirect_to trips_path
+    end
```
Comment (CheezItMan) [inline test/controllers/trips_controller_test.rb:126]: I think you intended to try to give a rating to an invalid trip. 

```suggestion
    it "will redirect to the trips page if given an invalid id" do
      # Your code here
      id = -1
      expect{
        patch assign_rating_trip_path(id), params: {rating: 4}
      }.wont_change "Trip.count"
      
      must_redirect_to trips_path
    end
```

File: test/controllers/trips_controller_test.rb [hunk 3]
```diff
@@ -1,23 +1,153 @@
...
+
+    it "will respond with redirect when attempting to edit a nonexistant trip" do
+      get edit_trip_path(-1)
+
+      must_respond_with :redirect
+    end
   end
 
   describe "update" do
```
Comment (CheezItMan) [inline test/controllers/trips_controller_test.rb:61]: What about if the user submits an update, but violates a validation?

File: test/controllers/trips_controller_test.rb [hunk 4]
```diff
@@ -1,23 +1,153 @@
...
+      get trip_path(-1)
+
+      # Assert
+      must_respond_with :redirect
+    end
   end
 
   describe "create" do
     # Your tests go here
+    it "can create a new trip" do
```
Comment (CheezItMan) [inline test/controllers/trips_controller_test.rb:36]: What about creating an invalid trip?

File: test/controllers/passengers_controller_test.rb [hunk 5]
```diff
@@ -1,31 +1,154 @@
...
+    end
+
+    it "will respond with redirect when attempting to edit a nonexistant passenger" do
+      get edit_passenger_path(-1)
+
+      must_respond_with :redirect
+    end
   end
 
   describe "update" do
```
Comment (CheezItMan) [inline test/controllers/passengers_controller_test.rb:89]: What about updating a passenger with invalid data?

File: test/controllers/passengers_controller_test.rb [hunk 6]
```diff
@@ -1,31 +1,154 @@
...
+      get new_passenger_path
+
+      # Assert
+      must_respond_with :success
+    end
   end
 
   describe "create" do
     # Your tests go here
+    it "can create a new passenger" do
```
Comment (CheezItMan) [inline test/controllers/passengers_controller_test.rb:51]: What about creating a passenger with invalid data?

File: test/controllers/drivers_controller_test.rb [hunk 7]
```diff
@@ -55,121 +51,183 @@
...
 
       # Act-Assert
       # Ensure that there is no change in Driver.count
+      expect {
+        post drivers_path, params: driver_hash
+      }.wont_change "Driver.count"
 
-      # Assert
-      # Check that the controller redirects
+      assert_template :new
```
Comment (CheezItMan) [inline test/controllers/drivers_controller_test.rb:192]: You don't need to test which template shows (this is depreciated in Rails nowadays).  

You **should** check the status code.

File: test/controllers/drivers_controller_test.rb [hunk 8]
```diff
@@ -55,121 +51,183 @@
...
```
Comment (CheezItMan) [inline test/controllers/drivers_controller_test.rb:97]: Better to just check the status code.  

```suggestion
      must_respond_with :bad_request
```

File: app/controllers/drivers_controller.rb [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+    @driver = Driver.new(
+        name: params[:driver][:name],
+        vin: params[:driver][:vin],
+        available: true
+    )
+    if @driver.save
+      redirect_to driver_path(@driver)
+      return
+    else
+      render :new
```
Comment (CheezItMan) [inline app/controllers/drivers_controller.rb:29]: You should also include a status code. 

```suggestion
      render :new, status: :bad_request
```

File: app/controllers/drivers_controller.rb [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+    result = @driver.update({
+                                name: params[:driver][:name],
+                                vin: params[:driver][:vin],
+                                available: params[:driver][:available]
+                          })
+
+    if result
+      redirect_to driver_path(@driver.id)
+    else
+      render :edit
```
Comment (CheezItMan) [inline app/controllers/drivers_controller.rb:54]:
```suggestion
      render :edit, status: :bad_request
```

File: app/controllers/passengers_controller.rb [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+
+  def show
+    @passenger = find_passenger
+    redirect_to passengers_path and return if @passenger.nil?
+  end
+
+  def create
+    @passenger = Passenger.create(passenger_params)
+    if @passenger.id.nil?
+      render :new
```
Comment (CheezItMan) [inline app/controllers/passengers_controller.rb:15]: Just a suggestion for readability?

```suggestion
    unless @passenger.valid?
      render :new, status: :bad_request
```

File: app/controllers/passengers_controller.rb [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+
+    redirect_to passengers_path and return if @passenger.nil?
+  end
+
+  def update
+    @passenger = find_passenger
+
+    redirect_to passengers_path and return if @passenger.nil?
+
+    action_success_check(@passenger.update(passenger_params), passenger_path, destination_view: :edit)
```
Comment (CheezItMan) [inline app/controllers/passengers_controller.rb:38]: interesting helper method.

File: app/views/passengers/edit.html.erb [hunk 13]
```diff
@@ -0,0 +1,27 @@
+<h2>Edit Passenger</h2>
```
Comment (CheezItMan) [inline app/views/passengers/edit.html.erb:1]: Why no partial?

File: app/views/passengers/show.html.erb [hunk 14]
```diff
@@ -0,0 +1,19 @@
+<h2>Passenger Details</h2>
+
+<div><strong>Passenger: <%= @passenger.name %></strong>
+  <p><strong>Phone number: </strong><%= @passenger.phone_num %></p>
+  <p><strong>Total Amount Charged: $</strong><%= @passenger.total_charged %></p>
+  <strong>Passenger's Trips:</strong>
+  <ol>
+    <% @passenger.trips.each do |trip| %>
+      <li>Trip ID:<%= link_to trip.id, trip_path(trip.id) %>, driver: <%= link_to trip.driver.name, driver_path(trip.driver.id) %>, Date: <%= trip.date %>, rating: <%= trip.rating %>, cost: $<%= trip.money %></li>
+    <%end %>
```
Comment (CheezItMan) [inline app/views/passengers/show.html.erb:10]: This can be done with a partial so you can re-use it any time you need a list of trips.

File: app/views/trips/edit.html.erb [hunk 15]
```diff
@@ -0,0 +1,28 @@
+<h2>Edit Trips</h2>
+
+<% if @trip.errors.any? %>
```
Comment (CheezItMan) [inline app/views/trips/edit.html.erb:3]: Again, this is a good candidate for a partial.

File: app/views/trips/show.html.erb [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+  <p><strong>Passenger:</strong> <%= link_to @trip.passenger.name, driver_path(@trip.passenger.id) %></p>
+  <p><strong>Cost:</strong> $<%='%.2f' % @trip.money %></p>
+  <p><strong>Rating:</strong> <%= @trip.rating %></p>
+  <%= form_for @trip do |f| %>
+    <%= f.select :rating, [['1', 1], ['2', 2], ['3', 3], ['4', 4], ['5', 5]] %>
+    <br>
+    <br>
+    <%= f.submit %>
+
+  <% end %>
```
Comment (CheezItMan) [inline app/views/trips/show.html.erb:15]: I like this!

File: app/models/driver.rb [hunk 17]
```diff
@@ -0,0 +1,21 @@
+class Driver < ApplicationRecord
+  has_many :trips
+
+  validates :name, presence: true
+  validates :vin, presence: true , length: { is: 17 }
+
+  def toggle_available
+    self.update(available: !self.available)
+  end
```
Comment (CheezItMan) [inline app/models/driver.rb:9]: 👍

File: app/models/driver.rb [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+
+  def total_earnings
+    gross_earnings = self.trips.sum{|trip| trip.cost - 165}
+
+    return gross_earnings > 165 ? ((gross_earnings) * 0.8).round : 0
+  end
+
+  def avg_rating
+    return self.trips.empty? ? 0 : (self.trips.map{|trip| trip.rating}.compact.sum/self.trips.length).round
+  end
```
Comment (CheezItMan) [inline app/models/driver.rb:19]: Great compact helper methods.

File: app/models/trip.rb [hunk 19]
```diff
@@ -0,0 +1,15 @@
...
+
+  validates :date, presence: true
+
+  validates :rating, numericality: { only_integer: true, greater_than: 0,less_than: 6 }, allow_nil: true
+  validates :cost, numericality: { only_integer: true, greater_than: 0 }
+
+  def money
+    in_dollars =  self[:cost]/ 100.00
+    return '%.2f' % in_dollars
+  end
```
Comment (CheezItMan) [inline app/models/trip.rb:13]: Just a note, this might be better as a view helper (more on that later)

File: config/routes.rb [hunk 20]
```diff
@@ -0,0 +1,17 @@
+Rails.application.routes.draw do
+  # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
+  resources :drivers
+  resources :passengers
+  resources :homepages, only: [:index]
+  resources :trips, except: [:create] do
+    member do
+      patch :assign_rating
+    end
+  end
```
Comment (CheezItMan) [inline config/routes.rb:10]: Neat!

File: config/routes.rb [hunk 21]
```diff
@@ -0,0 +1,17 @@
...
+  resources :passengers do
+    resources :trips, only: [:index, :create]
```
Comment (CheezItMan) [inline config/routes.rb:12]: I don't see you using the passenger_trips route?

File: config/routes.rb [hunk 22]
```diff
@@ -0,0 +1,17 @@
...
+  end
+  delete '/drivers/:id', to: 'drivers#delete', as: 'delete_driver'
```
Comment (CheezItMan) [inline config/routes.rb:14]: Do you need this?  You have resources.

File: test/models/passenger_test.rb [hunk 23]
```diff
@@ -61,12 +61,23 @@
...
+      it "returns false if there is no available driver" do
+        Trip.create(date: Date.current, cost: 7777, driver: Driver.new(name: "Kari", vin: "WBWSS52P9NEYLVDE9", available: false), passenger: new_passenger)
+        expect(new_passenger.request_trip).must_equal false
+      end
     end
 
-    describe "complete trip" do
-      # Your code here
+    describe "total_charged" do
+      it "correctly displays total charged"do
```
Comment (CheezItMan) [inline test/models/passenger_test.rb:76]: What about a passenger with no trips?

File: test/models/driver_test.rb [hunk 24]
```diff
@@ -54,28 +54,66 @@
...
+        expected_avg_rating = (ratings.sum/ratings.length).round
+
+        new_passenger = Passenger.create(name: "Kari", phone_num: "111-111-1211")
+        ratings.each{|rating| Trip.create!(rating: rating, cost:7777, date: Date.current, driver:new_driver, passenger: new_passenger)}
+
+        expect(new_driver.avg_rating).must_equal expected_avg_rating
+      end
     end
 
     describe "total earnings" do
```
Comment (CheezItMan) [inline test/models/driver_test.rb:86]: What about a driver with no trips?

File: test/models/driver_test.rb [hunk 25]
```diff
@@ -54,28 +54,66 @@
...
+      expect(new_driver.valid?).must_equal false
+      expect(new_driver.errors.messages).must_include :vin
+      d = new_driver.errors.messages
+      expect(new_driver.errors.messages[:vin][0]).must_equal "is the wrong length (should be 17 characters)"
     end
   end
 
   # Tests for methods you create should go here
   describe "custom methods" do
     describe "average rating" do
```
Comment (CheezItMan) [inline test/models/driver_test.rb:74]: What about a driver with no trips?

File: app/controllers/drivers_controller.rb [hunk 26]
```diff
@@ -0,0 +1,83 @@
...
+
+  def update
+    @driver = Driver.find(params[:id])
+    result = @driver.update({
+                                name: params[:driver][:name],
+                                vin: params[:driver][:vin],
+                                available: params[:driver][:available]
+                          })
+
+    if result
```
Comment (CheezItMan) [inline app/controllers/drivers_controller.rb:51]:
```suggestion
    if result.valid?
```

**Review by CheezItMan — COMMENTED** [review]
# Rideshare Rails



## Major Learning Goals/Code Review





| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
| Practices collaborating with git, and all team members contribute git commits and best git practices | ✔️
| Demonstrates understanding of relationships by giving accurate answers to the reflection questions and appropriate code in the models | ✔️
| Logic to calculate a driver's average rating and total earnings is located in the model, and it has unit tests | ✔️, however you need to test some edge cases see my inline methods.
| There are reasonable tests to test the validation requirements for all models | ✔️
| There are reasonable tests to test the relationship requirements for all models | ✔️
| There are reasonable tests to test the controller actions for all controllers | ✔️
| The app has an attractive and usable user interface | ✔️, although I think things would look better in a table-like format.
| The app uses/is compatible with database seeds | ✔️
| Router code is clean: uses `resources` and RESTful routes | ✔️, however see some of my notes

## Functional Requirements

| Functional Requirement | yes/no |
| --- | --- |
| On the passenger's details page, I want to be able to see total charged, list of trips, a link to edit, and delete | ✔️
| When adding a new passenger, I want to see errors and validations that show that a passenger must be provided a name and a phone number, so that I cannot make a passenger without name or phone number | ✔️
| On the passenger's details page, I can create a new trip for this passenger, with an assigned driver and no rating | ✔️
| On the driver's details page, I want to be able to see total earnings, average rating, list of trips, a link to edit, and delete | ✔️
| When adding a new driver, I want to see errors and validations that show that a driver must be provided a name and VIN, so that I cannot make a driver without name or VIN | ✔️
| On the trip's detail page, I want to be able to view details, assign a rating, navigate to the trip's passenger, driver, a link to edit, and delete | ✔️

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 8+ in Code Review && 5+ in Functional Requirements | 






## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Elegant/Clever | ✅
| Descriptive/Readable | ✅
| Concise | ✅
| Logical/Organized | ✅

## Summary

Really well done.  You hit the learning goals here.  I like some of your helper methods and you have some good tests too.  However you needed to test some edge cases on the custom methods, and I found a few smaller things to make note of, status codes, etc.  Take a look and let me know what questions you have.
